File: terminale/algorithmique/diviser-pour-regner/fonction-sorted/scripts/tri_rapide.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
@Author: Christophe Viroulaud
@Time:   2021/06/22 22:56:11

version tri en place
"""

from random import randint


def partitionner(tab: list, deb: int, fin: int) -> None:
    pivot = tab[deb]
    pos = deb
    for i in range(deb+1, fin+1):
        if tab[i] < pivot:
            pos += 1
            tab[i], tab[pos] = tab[pos], tab[i]
    # place le pivot
    tab[deb], tab[pos] = tab[pos], tab[deb]
    return pos


def tri_rapide(tab: list, deb: int, fin: int) -> None:
    if deb < fin:
        pivot = partitionner(tab, deb, fin)
        tri_rapide(tab, deb, pivot-1)
        tri_rapide(tab, pivot+1, fin)


t = [randint(0, 100) for _ in range(10)]
print(t)
tri_rapide(t, 0, len(t)-1)
print(t)

File: terminale/algorithmique/diviser-pour-regner/fonction-sorted/scripts/ocaml/tri_rapide.ml
Random.self_init();;

let affiche t = let j = 0 in 
                      let rec affiche_rec t i = if i<Array.length t then 
                              begin print_int t.(i);
                                    print_string " ";
                                    affiche_rec t (i+1) end 
                                  in affiche_rec t j;print_string "\n";;

print_string "tri rapide\n";;
let tab = Array.init 10 (fun i -> Random.int 100);;
affiche tab;;

let tri_rapide t = let partitionner deb fin = 